linalg_ops/attention/generate_e2e_attention_tests.py

```diff
@@ -0,0 +1,503 @@
+    if shapes_id == ShapesId.SMALL:
+        return [
+            TestShapeAndScale(batch=2, m=256, k1=64, k2=32, n=16, scale=1.0),
+        ]
+    if shapes_id == ShapesId.MEDIUM:
+        return [
+            TestShapeAndScale(batch=2, m=512, k1=128, k2=64, n=32, scale=1.0),
+        ]
+    if shapes_id == ShapesId.LARGE:
+        return [
+            TestShapeAndScale(batch=2, m=1024, k1=128, k2=128, n=64, scale=1.0),
+        ]
+
+    raise ValueError(shapes_id)
+
+
+# Determines the shape of input and kernel tensors.
+@dataclasses.dataclass
+class TestInputTensorShapes:
+    batch: int
+    m: int
+    k1: int
+    k2: int
+    n: int
+    scale: float
+
+
+# Helper for generate_function. Generates TestInputTensorShapes, i.e.
+# converts from the runtime shape dimensions in TestShape and given dynamicity to
+# the set of shapes to be used in a test function's input tensors.
+def generate_shapes_and_scale(shape: TestShapeAndScale):
+    batch = shape.batch
+    m = shape.m
+    k1 = shape.k1
+    k2 = shape.k2
+    n = shape.n
+    scale = shape.scale
+
+    shapes_scale = TestInputTensorShapes(
+        batch=batch,
+        m=m,
+        k1=k1,
+        k2=k2,
+        n=n,
+        scale=scale,
+    )
+    return shapes_scale
+
+
+# Helper to return input, kernel and output shapes based on the layout and the Attention Params.
+def get_tensor_shapes(
+    shapes_scale: TestShapeAndScale,
+):
+    batch = shapes_scale.batch
+    m = shapes_scale.m
+    k1 = shapes_scale.k1
+    k2 = shapes_scale.k2
+    n = shapes_scale.n
+    scale = shapes_scale.scale
+
+    query_tensor_shape = [batch, m, k1]
+    key_tensor_shape = [batch, k2, k1]
+    value_tensor_shape = [batch, k2, n]
+    result_tensor_shape = [batch, m, n]
+
+    return query_tensor_shape, key_tensor_shape, value_tensor_shape, result_tensor_shape
+
+
+# Helper for generate_function.
+# Generates a name for a test function in the generated MLIR code.
+def generate_function_name(
+    query_type: QueryElemTypeId,
+    key_type: KeyElemTypeId,
+    value_type: ValueElemTypeId,
+    shapes_scale: TestInputTensorShapes,
+):
+    query_t = query_type.value
+    key_t = key_type.value
+    value_t = value_type.value
+    result_t = value_type.value
+
+    batch = shapes_scale.batch
+    m = shapes_scale.m
+    k1 = shapes_scale.k1
+    k2 = shapes_scale.k2
+    n = shapes_scale.n
+
+    attention = "attention"
+    return (
+        f"{attention}_{batch}_{m}_{k1}_{k2}_{n}"
+        + f"_dtype_{query_t}_{key_t}_{value_t}_{result_t}"
+    )
+
+
+# Represents a generated test function.
+@dataclasses.dataclass
+class MLIRFunction:
+    name: str
+    signature: str
+    import_declaration: str
+    definition: str
+
+
+# Generates a test function in the generated MLIR code.
+# The generated function will take the same arguments as iree_linalg_ext.attention variants
+# and will just call iree_linalg_ext.attention variants with them, returning its result.
+def generate_function(
```

Comment:
Bringing back some discussion from https://github.com/iree-org/iree/pull/17751#discussion_r1714405252, could this generator produce operations other than `iree_linalg_ext.attention`? Fine to start with only that one, but what about `iree_linalg_ext.online_attention` or even `torch.nn.functional.scaled_dot_product_attention`?

Reply:
Better to ask this Kunwar @Groverkss.

Reply:
> Bringing back some discussion from [iree-org/iree#17751 (comment)](https://github.com/iree-org/iree/pull/17751#discussion_r1714405252), could this generator produce operations other than `iree_linalg_ext.attention`? Fine to start with only that one, but what about `iree_linalg_ext.online_attention` or even `torch.nn.functional.scaled_dot_product_attention`?

iree_linalg_ext.attention is really a frontend op. It's entire purpose is for frontends to have a drop in op to lower to for attention. This op is what every frontend (unless they know better) should lower to for attention. This is why testing `torch.nn.functional.scaled_dot_product_attention` doesn't make much sense to me. That operation should always lower to `iree_linalg_ext.attention`. `iree_linalg_ext.attention` also has broader support for attention variants than the torch op. Maybe at some point it could go live in `iree_input` also.

`iree_linalg_ext.online_attention` is a non-standard, compiler operation. It's not defined by any literature and the definition can change to whatever is the best thing to codegen. An input shouldn't really contain it unless they really know what they are doing. I don't think testing it e2e makes much sense, since you aren't expected to use it in input IR.

Reply:
Thanks, Kunwar, for the clarification.